pkg/kubelet/cm/dra/claiminfo.go

```diff
@@ -161,36 +153,76 @@ func newClaimInfoCache(stateDir, checkpointName string) (*claimInfoCache, error)
 			entry.ResourceHandles,
 		)
 		for pluginName, cdiDevices := range entry.CDIDevices {
-			err := info.addCDIDevices(pluginName, cdiDevices)
+			err := info.updateCDIDevices(pluginName, cdiDevices)
 			if err != nil {
 				return nil, fmt.Errorf("failed to add CDIDevices to claimInfo %+v: %+v", info, err)
 			}
 		}
-		cache.add(info)
+		if err = cache.add(&info); err != nil {
+			return nil, fmt.Errorf("failed to add claimInfo %+v to the cache: %w", info, err)
+		}
 	}
 
 	return cache, nil
 }
 
-func (cache *claimInfoCache) add(res *ClaimInfo) {
+func (cache *claimInfoCache) add(claimInfo *ClaimInfo) error {
 	cache.Lock()
 	defer cache.Unlock()
 
-	cache.claimInfo[res.ClaimName+res.Namespace] = res
+	if _, ok := cache.claimInfo[claimInfo.ClaimName+claimInfo.Namespace]; ok {
+		return fmt.Errorf("claim %s, namespace %s: claim info already in cache", claimInfo.ClaimName, claimInfo.Namespace)
+	}
+
+	cache.claimInfo[claimInfo.ClaimName+claimInfo.Namespace] = claimInfo.DeepCopy()
+	return nil
 }
 
-func (cache *claimInfoCache) get(claimName, namespace string) *ClaimInfo {
+func (cache *claimInfoCache) update(claimName, namespace string, pluginName string, cdiDevices []string) error {
+	cache.Lock()
+	defer cache.Unlock()
+
+	claimInfo, ok := cache.claimInfo[claimName+namespace]
+	if !ok {
+		return fmt.Errorf("claim %s, namespace %s: claim info not in cache", claimName, namespace)
+	}
+
+	claimInfo.prepared = true
```

Comment:
It feels awkward to be setting this in here

Reply:
Having more calls would mean keeping cache unlocked between the calls. I tried to avoid doing that. BTW, this is why I called this method `update`. It not only updates CDI devices, but also sets `prepared` field and will possibly do other updates in future.

Reply:
I feel pretty strongly that this shouldn't be in here -- especially since we end up calling this once for each pluginName, essentially overwriting this to `true` each time.

Reply:
If it's not here then this UnprepareResources code
```
			if err = m.cache.update(claimInfo.ClaimName, pod.Namespace, pluginName, result.GetCDIDevices()); err != nil {
				return fmt.Errorf("failed to update claimInfo cache: claim info: %+v, error: %w", claimInfo, err)
			}
```
would look like this:
```
			if err = m.cache.addCDIDevices(claimInfo.ClaimName, pod.Namespace, pluginName, result.GetCDIDevices()); err != nil {
				return fmt.Errorf("failed to update claimInfo cache: claim info: %+v, error: %w", claimInfo, err)
			}
                         // cache is unlocked here and this claim info can be changed from another goroutine
                         if err = m.cache.setPrepared(claimInfo.ClaimName, pod.Namespace, true); err != nil {
                         }
```
So, it lefts cache unlocked in the middle of the update. I intentionally made update as one function to avoid that.

Reply:
If it should be considered as prepared only after the loop, we should probably change GRPC API to accept slice of claims and call it only once.

Reply:
We can't each call goes to a different plugin.

Reply:
That's a good point. We should update the cache after all calls succeed, but this actually means that we should update cache inside the loop either.

Reply:
I mean even after the loop it would still be better to update the cache with one lock(`m.cache.update`) than with two (`m.cache.addCDIDevices` and `m.cache.setPrepared`), no?

Reply:
> We can't -- each call goes to a different plugin.

Yes, but resource shouldn't be considered Prepared if one of the plugins fails to prepare it.